packages/contracts/contracts/libs/v0.8.x/minter-libs/SettlementExpLib.sol

```diff
@@ -0,0 +1,381 @@
+        address _coreContract,
+        SettlementAuctionProjectConfig storage _settlementAuctionProjectConfig,
+        MaxInvocationsLib.MaxInvocationsProjectConfig
+            storage _maxInvocationsProjectConfig,
+        DAExpLib.DAProjectConfig storage _DAProjectConfig
+    ) internal returns (uint256 netRevenues) {
+        // require revenues to not have already been collected
+        require(
+            !_settlementAuctionProjectConfig.auctionRevenuesCollected,
+            "Revenues already collected"
+        );
+        // refresh max invocations, updating any local values that are
+        // illogical with respect to the current core contract state, and
+        // ensuring that local hasMaxBeenInvoked is accurate.
+        MaxInvocationsLib.refreshMaxInvocations(
+            _projectId,
+            _coreContract,
+            _maxInvocationsProjectConfig
+        );
+
+        // get the current net price of the auction - reverts if no auction
+        // is configured.
+        // @dev we use _getPriceUnsafe here, since we just safely synced the
+        // project's max invocations and maxHasBeenInvoked, which guarantees
+        // an accurate price calculation from _getPriceUnsafe, while being
+        // more gas efficient than _getPriceSafe.
+        // @dev price is guaranteed <= _projectConfig.latestPurchasePrice,
+        // since this minter enforces monotonically decreasing purchase prices.
+        uint256 _price = getPriceUnsafe({
+            _settlementAuctionProjectConfig: _settlementAuctionProjectConfig,
+            _maxInvocationsProjectConfig: _maxInvocationsProjectConfig,
+            _DAProjectConfig: _DAProjectConfig
+        });
+        // if the price is not base price, require that the auction have
+        // reached max invocations. This prevents premature withdrawl
+        // before final auction price is possible to know.
+        if (_price != _DAProjectConfig.basePrice) {
+            // @dev we can trust maxHasBeenInvoked, since we just
+            // refreshed it above with _refreshMaxInvocations, preventing any
+            // false negatives
+            require(
+                _maxInvocationsProjectConfig.maxHasBeenInvoked,
+                "Active auction not yet sold out"
+            );
+        } else {
+            uint128 basePrice = _DAProjectConfig.basePrice;
+            // base price of zero indicates no sales, since base price of zero
+            // is not allowed when configuring an auction.
+            // @dev no coverage else branch of following line because redundant
+            require(basePrice > 0, "Only latestPurchasePrice > 0");
+            // if the price is base price, the auction is valid and may be claimed
+            // update the latest purchase price to the base price, to ensure
+            // the base price is used for all future settlement calculations
+            _settlementAuctionProjectConfig.latestPurchasePrice = basePrice;
```

Comment:
nit: this is an effect before // EFFECTS below

Reply:
moved up in [5ea7ad3](https://github.com/ArtBlocks/artblocks-contracts/pull/928/commits/5ea7ad35091928e5318f5eb6cf156cae266e8336)